Fix bug with namespace reservation and snupkg push with older clients

shishirx34 opened the pull request:
Addresses https://github.com/NuGet/NuGetGallery/issues/6647 and https://github.com/NuGet/Engineering/issues/1983. Also, added tests.

![image](https://user-images.githubusercontent.com/1646506/50126339-6436cc80-0221-11e9-88ca-f5299360f9c5.png)

![image](https://user-images.githubusercontent.com/1646506/50188816-b4706600-02d7-11e9-9277-8d17c8f4a18c.png)

skofman1 commented:
Not sure regarding the error message for snupkgs. You could use a new client but in the push command use https://nuget.org/api/v2/package and get the same result. 

How about:
"You are trying to upload a symbols package (snupkg) as a package (nupkg). For details on how to upload symbols see https://docs.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg#publishing-a-symbol-package"

@anangaur ?

anangaur commented:
I dont know the context. What is the error condition? Is it for older clients or newer clients trying to upload snupkg as nupkg?

skofman1 commented:
The error condition is when a someone is trying to upload a snupkg using nupkg API

anangaur commented:
So someone is using the following?
```
> dotnet nuget push My.Sample.Pakacge.snupkg
```
This seems perfectly fine to me.

skofman1 commented:
@anangaur , it's fine if this is a new client that can distinguish snupkg from nupkg. If it's an old client it will break.

anangaur commented:
So essentially this is a case of old client not supporting push of snupkg. If that's the case, I like the earlier error message better:
https://user-images.githubusercontent.com/1646506/50126353-7153bb80-0221-11e9-8a2a-27a58bfc975c.png

This would helps users to unblock themselves without referring to the documentation.

anangaur commented:
IMO `dotnet nuget push My.Sample.Pakacge.snupkg -source https://nuget.org/api/v2/package` should just work! Can we enable this?

shishirx34 commented:
@joelverhagen - we haven't done any prod data scan for such packages yet. It was discovered before we went live with the symbols package, so I doubt there are any packages which might have symbols(apart from some test ones that @karann-msft added), I'll run the scan nevertheless.

I think the reflow should just work fine, it does not perform the initial package validations again and our backend pipelines are separate for nupkg and snupkg. I will test it out nonetheless.

agr commented on src/NuGetGallery/Services/ReservedNamespaceService.cs:
```diff
@@ -15,7 +15,7 @@ namespace NuGetGallery
     public class ReservedNamespaceService : IReservedNamespaceService
     {
-        private static readonly Regex NamespaceRegex = new Regex(@"^\w+([_.-]\w+)*[.]?$", RegexOptions.Compiled | RegexOptions.ExplicitCapture);
+        private static readonly Regex NamespaceRegex = new Regex(@"^\w+([_.-]\w+)*[.-]?$", RegexOptions.Compiled | RegexOptions.ExplicitCapture);
```
Please, specify match timeout for a Regex.

shishirx34 replied:
We set the timeout in `OwinStartup`
```
            // Setting time out for all RegEx objects. Noted in remarks at https://msdn.microsoft.com/en-us/library/system.text.regularexpressions.regex.matchtimeout%28v=vs.110%29.aspx
            AppDomain.CurrentDomain.SetData("REGEX_DEFAULT_MATCH_TIMEOUT", TimeSpan.FromSeconds(10));
```

anangaur commented on src/NuGetGallery/Strings.resx:
```diff
@@ -1085,4 +1085,7 @@ The {1} Team</value>
     <comment>{0} is the list of duplicate package authors</comment>
   </data>
+  <data name="UploadPackage_SymbolsPackageNotAllowed" xml:space="preserve">
+    <value>The symbols package (snupkg) upload is not allowed using older NuGet clients. Please use the latest (v4.9.1 or above) nuget client version to upload the symbols package.</value>
```
@skofman1  Can we start NUxxxx series for errors from service as well? Benefits: Searchable on the web and leads to a specific location on the documentation page. The contents of this page can be easily changed/improved anytime.

shishirx34 replied:
I agree, but I think that is an overall bigger change that needs better planning and design. 

For this change I would like to get some inputs on the text here.

anangaur replied:
Proposed message: 
`Snupkg upload failed. Please use latest NuGet clients (v 4.9.0 or above) and a V3 endpoint to push Symbol packages. For details, refer https://docs.microsoft.com/nuget/create-packages/symbol-packages-snupkg`

Can you also raise a PR to the [docs page](https://docs.microsoft.com/nuget/create-packages/symbol-packages-snupkg) to include the v3 source requirement as a Note?

shishirx34 replied:
added to https://github.com/NuGet/NuGetGallery/issues/6767

shishirx34 commented:
![image](https://user-images.githubusercontent.com/1646506/50188716-55aaec80-02d7-11e9-864a-358f6835fc35.png)

shishirx34 commented:
@joelverhagen - I ran my large scale analysis on ~72K packages that were uploaded over the last month and a half, and only **two** packages uploaded by @karann-msft were the ones that had symbols package type. I have deleted these packages for consistency. Also, I was able to reflow these packages before I deleted it.

joelverhagen commented:
Thanks @shishirx34 -- nice work.
